### ISSUE TITLE
What's the difference between last.pt&best.pt, and which one should i use when i test my custom data?

### ISSUE THREAD

**Issue by PromiseXu1** [ultralytics/yolov5#582]
I have tried your work of yolov3 on my custom data, and when i train on yolov5  i find that yolov5 is more excellent! I just train for 100 epochs, the map50 have reached at 80! And i will train more epochs to see if it could be better .  
## But **the question is that** when i finish the train, and start to test, which weight.pt should i choose? And what's the difference between last.pt and best.pt? 
 Could you help me? Thanks sincerely.

**Comment by glenn-jocher** [comment]
@PromiseXu1 last.pt are the weights from the last epoch of training. best.pt are the best weights recorded during training.

**Comment by PromiseXu1** [comment]
@glenn-jocher  ah it seems so obviously, thanks for your patient... but i find sometimes best.pt not always be there, and i find someone have mentioned it, are you dealing with it?

**Comment by PromiseXu1** [comment]
![1](https://user-images.githubusercontent.com/56194924/89011247-9d627880-d342-11ea-9361-bf7af84c0ce2.png)

**Comment by TaoXieSZ** [comment]
@PromiseXu1 Because in exp11, when you resume from another exp, the model you trained didn't have better mAP than the one you resume from.

**Comment by glenn-jocher** [comment]
@PromiseXu1 --resuming still has a few bugs to iron out, so I would avoid it as much as possible. 

Best.pt will be saved anytime a new best fitness is observed on your validation set. We define fitness as a weighted combination of metrics:
https://github.com/ultralytics/yolov5/blob/48e15be498ab6f3335743b5df7fbd02edd068160/utils/utils.py#L881-L885

**Comment by Yu-Hang** [comment]
https://github.com/ultralytics/yolov5/blob/48e15be498ab6f3335743b5df7fbd02edd068160/utils/utils.py#L881-L885
Does this mean the weights for new best fitness will be recorded even if the model already overfits?

**Comment by glenn-jocher** [comment]
@Yu-Hang they are exactly as they sound, the last and best checkpoints during your training.

If the model overfits then by definition there will be no new best fitness.

**Comment by Yu-Hang** [comment]
Let me ask more specifically. The fitness function does not include any loss metrics. So, if you train for a large epoch, [P, R, mAP@0.5, mAP@0.5:0.95] are increasing, but some val loss starts to increases as well (which indicates overfitting?). In this case, will this creates new best fitness?

**Comment by glenn-jocher** [comment]
@Yu-Hang fitness function is defined here. best.pt is saved on every new best-fitness epoch.
https://github.com/ultralytics/yolov5/blob/417a2f425cd553c1f2a41374a6cd2710aa91d5f0/utils/metrics.py#L13-L17

**Comment by Petros626** [comment]
@glenn-jocher `best.pt` is used for the post training validation? These results are based on it?
